### ISSUE TITLE
span.context.destination.service.name is not set for forward span when using tracing to ElasticSearch

### ISSUE THREAD

**Issue by bmarkowww** [traefik/traefik#8158]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

We have thi setup:

1. Traefik and NodeJS app behind it
2. With enabled tracing to ElasticSearch

### What did you expect to see?

We expect to see distributed tracing visialization in kibana with traefik and node js app on the same timeline

### What did you see instead?

Instead we could see tracing timeline for either traefik or nodejs app but not both simultaneously

![image](https://user-images.githubusercontent.com/12733578/119652592-f7cb4900-be2e-11eb-8733-646074e769bd.png)

I did a small research and found this:
![image](https://user-images.githubusercontent.com/12733578/119651192-6a3b2980-be2d-11eb-917b-2e52f3a6c0a2.png)

parent.id for transaction in imt (nodejs app) should be equal to traefik's transaction.id in order for Elastic Search to place spans in correct order. Other spans on screenshot have correct parent id which is ether transcation started by app or another span.

I manually edited one app transaction parent.id to match corresponding traefik's transcation.id and then Elastic correctly displayed distributed trace for that transaction
![image](https://user-images.githubusercontent.com/12733578/119652465-ce122200-be2e-11eb-9541-2e10015e9237.png)

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.4.8
Codename:     livarot
Go version:   go1.16.2
Built:        2021-03-23T15:48:39Z
OS/Arch:      linux/amd64

```

**Comment by tomMoulard** [comment]
Hello @bmarkowww,

Thanks for your interest for Traefik!

Could you please provide a reproducible case?

**Comment by bmarkowww** [comment]
Hi @tomMoulard,

Let's try to reproduce this case
First of all we need to run traefik, some upstream service (whoami in this case) and elastic stack (elasticsearch, kibana and apm-server)

**docker-compose.yml**

```
version: '2.2'

services:
  whoami:
    image: containous/whoami
    container_name: whoami
    restart: always
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.whoami-https.entrypoints=http"
      - "traefik.http.routers.whoami-https.rule=Host(`localhost`) && PathPrefix(`/whoami`)"
    networks:
    - elastic
  traefik:
    image: traefik:latest
    container_name: "traefik"
    restart: always
    ports:
      - "80:80"
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
      - "./traefik:/etc/traefik"
    environment:
      - ELASTIC_APM_CENTRAL_CONFIG=false
    networks:
    - elastic
  apm-server:
    image: docker.elastic.co/apm/apm-server:7.13.1
    container_name: apm-server
    depends_on:
      elasticsearch:
        condition: service_healthy
      kibana:
        condition: service_healthy
    cap_add: ["CHOWN", "DAC_OVERRIDE", "SETGID", "SETUID"]
    cap_drop: ["ALL"]
    ports:
    - 8200:8200
    networks:
    - elastic
    command: >
       apm-server -e
         -E apm-server.rum.enabled=true
         -E setup.kibana.host=kibana:5601
         -E setup.template.settings.index.number_of_replicas=0
         -E apm-server.kibana.enabled=true
         -E apm-server.kibana.host=kibana:5601
         -E output.elasticsearch.hosts=["elasticsearch:9200"]
    healthcheck:
      interval: 10s
      retries: 12
      test: curl --write-out 'HTTP %{http_code}' --fail --silent --output /dev/null http://localhost:8200/

  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch:7.13.1
    container_name: elasticsearch
    environment:
    - bootstrap.memory_lock=true
    - cluster.name=docker-cluster
    - cluster.routing.allocation.disk.threshold_enabled=false
    - discovery.type=single-node
    - ES_JAVA_OPTS=-XX:UseAVX=2 -Xms1g -Xmx1g
    ulimits:
      memlock:
        hard: -1
        soft: -1
    volumes:
    - esdata:/usr/share/elasticsearch/data
    ports:
    - 9200:9200
    networks:
    - elastic
    healthcheck:
      interval: 20s
      retries: 10
      test: curl -s http://localhost:9200/_cluster/health | grep -vq '"status":"red"'

  kibana:
    image: docker.elastic.co/kibana/kibana:7.13.1
    container_name: kibana
    depends_on:
      elasticsearch:
        condition: service_healthy
    environment:
      ELASTICSEARCH_URL: http://elasticsearch:9200
      ELASTICSEARCH_HOSTS: http://elasticsearch:9200
    ports:
    - 5601:5601
    networks:
    - elastic
    healthcheck:
      interval: 10s
      retries: 20
      test: curl --write-out 'HTTP %{http_code}' --fail --silent --output /dev/null http://localhost:5601/api/status

volumes:
  esdata:
    driver: local

networks:
  elastic:
    driver: bridge
```

**traefik.yml**
```
entryPoints:
  http:
    address: ":80"

api:
  dashboard: true

providers:
  docker:
    watch: true
    exposedbydefault: false

log:
  filePath: "/etc/traefik/traefik.log"
  level: DEBUG

accessLog:
  filePath: "/etc/traefik/access.log"

tracing:
  serviceName: traefik
  elastic:
    serverURL: http://apm-server:8200

```

Now we're ready to do some requests to whoami service

```
curl -lv http://localhost/whoami
*   Trying 127.0.0.1:80...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /whoami HTTP/1.1
> Host: localhost
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Length: 497
< Content-Type: text/plain; charset=utf-8
< Date: Mon, 07 Jun 2021 21:38:25 GMT
< 
Hostname: 533618549cc1
IP: 127.0.0.1
IP: 172.22.0.5
RemoteAddr: 172.22.0.6:47282
GET /whoami HTTP/1.1
Host: localhost
User-Agent: curl/7.68.0
Accept: */*
Accept-Encoding: gzip
Elastic-Apm-Traceparent: 00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01
Traceparent: 00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01
X-Forwarded-For: 172.22.0.1
X-Forwarded-Host: localhost
X-Forwarded-Port: 80
X-Forwarded-Proto: http
X-Forwarded-Server: f810d2b51a4b
X-Real-Ip: 172.22.0.1
```

Then we should take a look at traefik's debug logs:

```
time="2021-06-07T21:38:25Z" level=debug msg="Failed to extract the context: opentracing: SpanContext not found in Extract carrier" middlewareName=tracing middlewareType=TracingEntryPoint
time="2021-06-07T21:38:25Z" level=debug msg="vulcand/oxy/roundrobin/rr: begin ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/whoami\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Elastic-Apm-Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"User-Agent\":[\"curl/7.68.0\"],\"X-Forwarded-Host\":[\"localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"f810d2b51a4b\"],\"X-Real-Ip\":[\"172.22.0.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.22.0.1:58888\",\"RequestURI\":\"/whoami\",\"TLS\":null}"
time="2021-06-07T21:38:25Z" level=debug msg="vulcand/oxy/roundrobin/rr: Forwarding this request to URL" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/whoami\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Elastic-Apm-Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"User-Agent\":[\"curl/7.68.0\"],\"X-Forwarded-Host\":[\"localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"f810d2b51a4b\"],\"X-Real-Ip\":[\"172.22.0.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.22.0.1:58888\",\"RequestURI\":\"/whoami\",\"TLS\":null}" ForwardURL="http://172.22.0.5:80"
time="2021-06-07T21:38:25Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/whoami\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Elastic-Apm-Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"Traceparent\":[\"00-4a64c5ce3f755a646c3dc1b466a5cc66-14ba5a9a87fed790-01\"],\"User-Agent\":[\"curl/7.68.0\"],\"X-Forwarded-Host\":[\"localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"f810d2b51a4b\"],\"X-Real-Ip\":[\"172.22.0.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.22.0.1:58888\",\"RequestURI\":\"/whoami\",\"TLS\":null}"
time="2021-06-07T21:38:35Z" level=debug msg="request failed: request failed with 400 Bad Request: {\"accepted\":1,\"errors\":[{\"message\":\"validation error: span: context: destination: service: 'name' required\",\"document\":\"{\\\"span\\\":{\\\"duration\\\":0.657652,\\\"id\\\":\\\"14ba5a9a87fed790\\\",\\\"name\\\":\\\"forward whoami-traefik-test/whoami-https@docker\\\",\\\"timestamp\\\":1623101905027047,\\\"trace_id\\\":\\\"4a64c5ce3f755a646c3dc1b466a5cc66\\\",\\\"transaction_id\\\":\\\"4a64c5ce3f755a64\\\",\\\"type\\\":\\\"external\\\",\\\"context\\\":{\\\"destination\\\":{\\\"service\\\":{\\\"type\\\":\\\"external\\\"}},\\\"http\\\":{\\\"url\\\":\\\"http:///whoami\\\"},\\\"tags\\\":{\\\"http_host\\\":\\\"localhost\\\",\\\"http_status_code\\\":\\\"200\\\",\\\"span_kind\\\":\\\"client\\\",\\\"service_name\\\":\\\"whoami-traefik-test\\\",\\\"router_name\\\":\\\"whoami-https@docker\\\"}},\\\"parent_id\\\":\\\"4a64c5ce3f755a64\\\",\\\"subtype\\\":\\\"http\\\"}}\"}]} (next request in ~0s)"
```

There we have two interesting lines:

1. `Failed to extract the context: opentracing: SpanContext not found in Extract carrier" middlewareName=tracing middlewareType=TracingEntryPoint` - it looks similar to https://github.com/traefik/traefik/issues/7855
2. `request failed: request failed with 400 Bad Request: {\"accepted\":1,\"errors\":[{\"message\":\"validation error: span: context: destination: service: 'name' required`

I think the second one is the most important. Let's extract json from that log line:
```
{
  "span": {
    "duration": 0.657652,
    "id": "14ba5a9a87fed790",
    "name": "forward whoami-traefik-test/whoami-https@docker",
    "timestamp": 1623101905027047,
    "trace_id": "4a64c5ce3f755a646c3dc1b466a5cc66",
    "transaction_id": "4a64c5ce3f755a64",
    "type": "external",
    "context": {
      "destination": {
        "service": {
          "type": "external"
        }
      },
      "http": {
        "url": "http://whoami"
      },
      "tags": {
        "http_host": "localhost",
        "http_status_code": "200",
        "span_kind": "client",
        "service_name": "whoami-traefik-test",
        "router_name": "whoami-https@docker"
      }
    },
    "parent_id": "4a64c5ce3f755a64",
    "subtype": "http"
  }
}
```
This json is missing `span.context.destination.service.name` field just like it said in the log so elastic doesn't record forward span. That span is crucial because its id should be the parent id for a transaction that could be started by backend service behind the traefik. But it's missing so elastic couldn't restore inheritance order.

I did more research and found:

1.  `span.context.destination.service.name` is a requirement for Kibana's service map feature and was disscussed in  https://github.com/elastic/apm/issues/180
2. `span.context.destination.service.name` was introduced in elastic apm-go-agent v 1.7.0 https://github.com/elastic/apm-agent-go/pull/664
3. Elastic Apm Server started to reject spans without `span.context.destination.service.name` somewhere between 7.5.2 and 7.6.2 versions
4. Empty `span.context.destination.service.name` was addressed in https://github.com/elastic/apm-agent-go/pull/931 and landed to v1.12.0 of elastic apm-go-agent

So I see two different options:

1. Bump elastic apm-go-agent version to 1.12.0. It will certainly fix the issue but will leave Kibana's service map feature non-functional
2. Introduce some way to set `span.context.destination.service.name`. The problem here is that traefik doesn't know with which service name an upstream service is configured but these names should be equal in order to get service map working. So there should be a way to set that name in traefik's config for every service

**Comment by mrnonz** [comment]
@bmarkowww I faced this issue like you on traefix `2.4.9` with Elastic 7.13, So do have any workaround solution for this?

**Comment by bmarkowww** [comment]
> @bmarkowww I faced this issue like you on traefix `2.4.9` with Elastic 7.13, So do have any workaround solution for this?

Hi mrnonz, 
I see two options here:

1. You could bump elastic/apm-go-agent version to v1.12.0 and build custom version of traefik manually. It's quite easy.
2. Or you could wait until traefik's team do it themselves. As i can see v1.11.0 got into 2.5.0-rc2 version of traefik.

**Comment by vedard** [comment]
Hi,

I've also encountered this problem. Do you have an ETA when the `elastic/apm-go-agent` will be bump to v1.12.0?

Thanks!

**Comment by rtribotte** [comment]
Hello @bmarkowww,

Thanks for the reproduction case.

I tried it and I was able to confirm your issue, I can confirm that a validation error is returned from the APM server.

I bumped the `go.elastic.co/apm` package version from v1.11.0 to v1.12.0 and it suppressed the validation error seen in Traefik logs.

Thus, something is not clear for me, is this issue preventing you to use the tracing at all?
Apparently, the spans are validated even though there is a validation error:
```
DEBU[2021-08-24T12:13:23+02:00] request failed: request failed with 400 Bad Request: {"accepted":3,"errors":[{"message":"validation error: span: context: destination: service: 'name' required","document":...
```
Also, I was able to see them in Kibana.

Nonetheless, I will open a PR to bump `go.elastic.co/apm` package to get rid of that validation error.

**Comment by vedard** [comment]
Hello @rtribotte 

To clarify, tracing was working, but one transaction was missing (the one with the validation error). This was affecting distributed tracing (when you have multiple apps linked together) because Kibana was not able to draw the full graph (due to the missing transaction).

Thanks you for resolving this!

**Comment by traefiker** [comment]
Closed by #8399.
